### ISSUE TITLE
关于ZipDataset的问题，还有合成后是否需要clamp的问题

### ISSUE THREAD

**Issue by eternalgogi92** [PeterL1n/BackgroundMattingV2#121]
Q1:如果A和B的items都是排序好的，是不是ZipDataset([A,B])在每个epoch读出来的pair都是固定的？比如A=[a1,a2,a3], B=[b1,b2,b3]，那么每个epoch中a1始终和b1配对。这种配对生成方式，合成出来的样本是否缺少多样性呢？我自己简单写了一个代码尝试了下，发现配对是不改变的。
代码如下：
```
from dataset import ZipDataset
from torch.utils.data import Dataset, DataLoader

class Dataset(Dataset):
    def __init__(self, tag, max_len):
        self.ids = [f'{tag}_{i}' for i in range(max_len)]

    def __getitem__(self, idx):
        return self.ids[idx]

    def __len__(self):
        return len(self.ids)

if __name__ == '__main__':
    short_db = Dataset('short', 2)
    long_db = Dataset('long', 3)
    zip_db = ZipDataset([short_db, long_db])
    dl = DataLoader(zip_db, batch_size=1, shuffle=True, pin_memory=False, sampler=None, num_workers=1)
    for (si, li) in dl:
        print(si, li)
    print('-'*35)
    for (si, li) in dl:
        print(si, li)
```
结果如下：
```
('short_1',) ('long_1',)
('short_0',) ('long_2',)
('short_0',) ('long_0',)
-----------------------------------
('short_0',) ('long_2',)
('short_0',) ('long_0',)
('short_1',) ('long_1',)
```
Q2:在train_base还有train_refine中，使用了true_src = true_fgr * true_pha + true_src * (1 - true_pha)公式后，没有立刻使用clamp，是因为后面加了noise中做了clamp操作吗？
谢谢。

**Comment by PeterL1n** [comment]
`true_src = true_fgr * true_pha + true_src * (1 - true_pha)` 不需要 clamp，因为每一项都是在 0-1 之间，最终结果也会是 0-1。加噪点后需要 clamp 因为会超过 0-1。

ZipDataset 是固定的，DataLoader 会做 shuffle

**Comment by eternalgogi92** [comment]
> `true_src = true_fgr * true_pha + true_src * (1 - true_pha)` 不需要 clamp，因为每一项都是在 0-1 之间，最终结果也会是 0-1。加噪点后需要 clamp 因为会超过 0-1。
> 
> ZipDataset 是固定的，DataLoader 会做 shuffle

但是我发现dataloader shuffle完了，ZipDataset中配对关系是不会改变的，比如fg1和bg1在第一个epoch中是配对的，在第2个epoch仍旧是配对的，请问是否有必要打乱这种配对关系呢？

**Comment by eternalgogi92** [comment]
> `true_src = true_fgr * true_pha + true_src * (1 - true_pha)` 不需要 clamp，因为每一项都是在 0-1 之间，最终结果也会是 0-1。加噪点后需要 clamp 因为会超过 0-1。
> 
> ZipDataset 是固定的，DataLoader 会做 shuffle

你好，我更新了代码和结果哈，麻烦您再看下。可以看到short和long中每个item对应的关系是不受dataloader shuffle的影响的。

**Comment by PeterL1n** [comment]
Zip是对应的。假设 A 是 Matting Dataset 有 100 个元素，B 是背景有 1000 个。那么每个 Matting sample 会固定对应 10 个背景。你可以自己更改代码

**Comment by eternalgogi92** [comment]
> Zip是对应的。假设 A 是 Matting Dataset 有 100 个元素，B 是背景有 1000 个。那么每个 Matting sample 会固定对应 10 个背景。你可以自己更改代码

谢谢
